src/sinks/kafka/service.rs

```diff
@@ -78,12 +99,25 @@ impl Service<KafkaRequest> for KafkaService {
     type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
 
     fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Poll::Ready(Ok(()))
+        // The Kafka service is available if its producer queue is not full.
+        if self.queue_messages_current.load(Ordering::Relaxed) < self.queue_messages_max
+            && self.queue_bytes_current.load(Ordering::Relaxed) < self.queue_bytes_max
+        {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Pending
+        }
     }
 
     fn call(&mut self, request: KafkaRequest) -> Self::Future {
         let this = self.clone();
 
+        // Update state for the producer queue.
+        let raw_byte_size = request.body.len() + request.metadata.key.map_or(0, |x| x.len());
```

Comment:
My primary question with this change is if this is an accurate measurement of the producer's size for the message in its internal queue.

Reply:
Looks like at minimum this wouldn't include the size of any headers that might be set on the message, right? Would tapping into the [producer statistics for queue size/count](https://docs.rs/rdkafka/latest/rdkafka/statistics/struct.Statistics.html#structfield.msg_size) be more reliable than tracking it externally?

Reply:
I was able to avoid this altogether using a different approach.